src/backend.rs

```diff
@@ -843,17 +891,24 @@ impl State {
         // This implementation is not nearly optimal, but it should be sufficient in a system
         // with relatively few services.
         let mut result = Vec::new();
-        self.with_services(selectors, |service| {
+        self.with_services(selectors, |service, _| {
             result.push(service.borrow().as_service())
         });
         result
     }
 
     pub fn add_service_tags(&mut self, selectors: Vec<ServiceSelector>, tags: Vec<Id<TagId>>) -> usize {
         let mut result = 0;
-        self.with_services(selectors, |service| {
+
+        self.with_services(selectors, |service, store| {
             let service = service.borrow_mut();
             let mut tag_set = service.tags.borrow_mut();
+
+            if let Some(ref storage) = *store {
+                storage.add_tags(&service.id, &tags)
+                       .unwrap_or_else(|err| { error!("Storage add_tags error: {}", err); });
```

Comment:
Nit: Can you make it clearer that this log is issued by the Taxonomy back-end?

Also, a `target` is probably useful, because we already have many logs.